test/test_main.py

```diff
@@ -499,5 +477,193 @@ def test_logger(self):
+        }))
+
+        self.fake_script_path = self.hooks_dir
+        fake_script(self, 'install', 'exit 0')
+        state = self._simulate_event(EventSpec(InstallEvent, 'install', charm_config=charm_config))
+
+        # the script was called, *and*, the .on. was called
+        self.assertEqual(fake_script_calls(self), [['install', '']])
+        self.assertEqual(state['observed_event_types'], [InstallEvent])
+
+    def test_hook_and_dispatch_with_failing_hook(self):
+        self.stdout = self.stderr = tempfile.TemporaryFile()
+        self.addCleanup(self.stdout.close)
+
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+
+        old_path = self.fake_script_path
+        self.fake_script_path = self.hooks_dir
+        fake_script(self, 'install', 'exit 42')
+        event = EventSpec(InstallEvent, 'install', charm_config=charm_config)
+        with self.assertRaises(subprocess.CalledProcessError):
+            self._simulate_event(event)
+        self.fake_script_path = old_path
+
+        self.stdout.seek(0)
+        self.assertEqual(self.stdout.read(), b'')
+        calls = fake_script_calls(self)
+        self.assertEqual(len(calls), 1)
+        self.assertEqual(len(calls[0]), 4)
+        self.assertEqual(
+            calls[0][:3],
+            ['juju-log', '--log-level', 'WARNING']
+        )
+        self.assertRegex(calls[0][3], r'hook /\S+/install exited with status 42')
+
+    def test_hook_and_dispatch_but_hook_is_dispatch(self):
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+        event = EventSpec(InstallEvent, 'install', charm_config=charm_config)
+        hook_path = self.hooks_dir / 'install'
+        for ((rel, ind), path) in {
+                # relative and indirect
+                (True, True):   Path('../dispatch'),
+                # relative and direct
+                (True, False):  Path(self.charm_exec_path),
+                # absolute and direct
+                (False, False): (self.hooks_dir / self.charm_exec_path).resolve(),
+                # absolute and indirect
+                (False, True):  self.JUJU_CHARM_DIR / 'dispatch',
+        }.items():
+            msg = "{} (relative:{} indirect:{})".format(path, rel, ind)
+            # sanity check
+            self.assertEqual(path.is_absolute(), not rel, msg)
+            self.assertEqual(path.name == 'dispatch', ind, msg)
+            hook_path.symlink_to(path)
+
+            state = self._simulate_event(event)
+
+            # the .on. was only called once
+            self.assertEqual(state['observed_event_types'], [InstallEvent], msg)
+            self.assertEqual(state['on_install'], [InstallEvent], msg)
+
+            hook_path.unlink()
+
+
+# TODO: this does not work
+# class TestMainWithDispatchAsScript(TestMainWithDispatch):
+#     """Here dispatch is a script that execs the charm.py instead of a symlink.
```

Comment:
If dispatch was a shell script that invoked src/charm.py I guess the issue is that [argv[0]] isn't 'dispatch' so we don't realize we want to invoke the dispatch code path?